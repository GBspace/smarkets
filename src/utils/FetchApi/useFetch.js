import { useEffect, useState, useRef } from "react";
  const useFetch = (url, ref, initialValue) => {
    const [data, setData] = useState(initialValue);
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(true);
    var cors_api_url = 'https://cors-anywhere.herokuapp.com/';
    useEffect(() => {
      if(ref.current){
        fetch(cors_api_url + url, {
          method: 'GET',
          headers: {'Content-Type': 'application/x-www-form-urlencoded'}
        })
        .then(res => {
          if(res.ok) {
            return res.json();
          } else {
            throw new Error('Something went wrong');
          }
        })
        .then(json => { 
          setData(json)
        })
        .catch(err => {
          setError(err.message);
        }).finally(()=>setLoading(false));
      }
      return () => {
        ref.current = false;
      }
    }, [url,ref]);
    return {data,error,loading};
  }

export {useFetch};