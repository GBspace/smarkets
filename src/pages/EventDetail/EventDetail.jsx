import React, {useRef} from 'react'
import { useParams } from 'react-router-dom'
import {useFetch} from '../../utils/FetchApi/useFetch';

const EventDetail = () => {
    const isComponentMounted = useRef(true);
    const {id} = useParams();
    const {data,error,loading} = useFetch(`https://api.smarkets.com/v3/events/${id}/`,isComponentMounted,[]);
    return (
        <div data-testid="eventDetail">
            {loading && <div className='loading'>Loading...</div>}
            {(data?.events) && 
                    <div className='data-container'>
                        <div className='name'>Event Name: {data.events[0].name} </div>
                        <div className='desc'>Event Description: {data.events[0].description} </div>
                        <div className='startTime'>Start Time: {data.events[0].start_datetime} </div>
                    </div>
            }
            {error && <div className="error-container">{error}</div>}
        </div>
    )
}

export default EventDetail;