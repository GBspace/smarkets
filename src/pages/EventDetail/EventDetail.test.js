import EventDetail from './EventDetail';
import {shallow} from 'enzyme';
import { useParams } from 'react-router-dom';
import  {useFetch} from '../../utils/FetchApi/useFetch';
import * as React from "react";

jest.mock('react-router-dom', () => {
    const spy = jest.fn()
    return {
      ...jest.requireActual('react-router-dom'),
      useParams: spy
    }
  });
  jest.mock('../../utils/FetchApi/useFetch', ()=>{
    const spy = jest.fn();
    return {
        __esModule: true,
    ...jest.requireActual('../../utils/FetchApi/useFetch'),
    useFetch: spy,
    }
});

describe('EventDetailPage', ()=>{
    it('should display loading till data is fetched', ()=>{
        useParams.mockImplementationOnce(() => ({
            id: '1'
          }));
         
    useFetch.mockImplementationOnce(()=>({
        data: null, 
        loading: true,
        error: null
    }));
        const wrapper = shallow(<EventDetail/>);
        const loadingDiv = wrapper.find('.loading');
        expect(loadingDiv.text()).toBe("Loading...");
        });

    it('should display the event deatils for an id', ()=>{
            useParams.mockImplementationOnce(() => ({
                id: '1'
              }));
             
        useFetch.mockImplementationOnce(()=>({
            data: {
                events: [
                    {
                        name: 'IPL',
                        description: 'Cricket League',
                        start_datetime: 'April 2022'
                    }
                ]
            }, 
            loading: false,
            error: null
        }));
        const wrapper = shallow(<EventDetail/>);
        const loadingDiv = wrapper.find('.loading');
        const name = wrapper.find('.name');
        const desc = wrapper.find('.desc');
        const startTime = wrapper.find('.startTime');
        expect(name.text()).toBe("Event Name: IPL ");
        expect(desc.text()).toBe("Event Description: Cricket League ");
        expect(startTime.text()).toBe("Start Time: April 2022 ");
        expect((loadingDiv).exists()).toBe(false);
    });

    it('should display Error with a failing newwork request', ()=>{
        useParams.mockImplementationOnce(() => ({
            id: '1'
          }));
         
    useFetch.mockImplementationOnce(()=>({
        data: null, 
        loading: false,
        error: 'Something went wrong'
    }));
        const wrapper = shallow(<EventDetail/>);
        const errorDiv = wrapper.find('.error-container');
        expect(errorDiv.text()).toBe("Something went wrong");
        });

})