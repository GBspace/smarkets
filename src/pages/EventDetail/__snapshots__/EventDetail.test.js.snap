// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventDetailPage renders correctly when events are loading 1`] = `
<Styled(div)>
  <div
    data-testid="eventDetail"
  >
    <div
      className="loading"
    >
      Loading...
    </div>
  </div>
</Styled(div)>
`;

exports[`EventDetailPage renders correctly when events details are fetched 1`] = `
<Styled(div)>
  <div
    data-testid="eventDetail"
  >
    <div
      className="data-container"
    >
      <div
        className="name"
      >
        <strong>
          Event Name: 
        </strong>
        IPL
         
      </div>
      <div
        className="desc"
      >
        <strong>
          Event Description:
        </strong>
         
        Cricket League
         
      </div>
      <div
        className="startTime"
      >
        <strong>
          Start Time:
        </strong>
        April 2022
         
      </div>
    </div>
  </div>
</Styled(div)>
`;
